### ISSUE TITLE
Step Plugins

### ISSUE THREAD

**Issue by zachaller** [argoproj/argo-rollouts#2685]
# Summary

Place holder for step plugins.

# Use Cases

When would you use this?

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by BasJ93** [comment]
I would be very interested in having the ability to use step plugins. In my particular use case I would benefit greatly from a step plugin that can function as a gate based on a metric rule.

My application is stateful in the sense that it sense that it maintains large in memory objects that are flushed to disk when certain conditions are met. The application exposes information about the state of these objects as metrics. Based on the metrics I would like the ability to determine if Rollouts is allowed to progress in to a step and replace one or more containers.

In short: I would love the ability to gate a step based on metrics. The step would wait until the metrics query returns a positive result and then continue with the step.

**Comment by GTB3NW** [comment]
Great talk!

I think a great step would be for chatops like you mentioned. I can foresee a template of steps which can then be applied and include their current status. So define your steps like normal but inside of _slack for example_ it would appear as a customizable template where you can click progress and have a live update as it progresses to the next step. A single point to track and progress the rollout

**Comment by akorzy-pl** [comment]
Hi, it would be useful to have a custom step that waits until a different rollout in a different cluster reaches the same step index. That would allow us to have more than one rollout happen simultaneously and in a synchronized fashion. The rollouts are going to need something like a common multiClusterRolloutId, so that the step can identify the correct rollout.

**Comment by akorzy-pl** [comment]
Hi again, it would be useful to have the ability to pass data between different steps of a rollout. For example, when we start a verification test, we need to call an API, which returns a test run ID. We need to use that ID in the next step, which will periodically poll the API and wait for the test results to become available. Currently we have no simple way to pass such an ID between steps in a rollout. In addition, we would like Argo Rollouts to preserve a history of previously passed data, so that we can debug previous rollouts when necessary.

**Comment by akorzy-pl** [comment]
Hello, we would like to have a custom step that allows us to send a notification to Slack. We don't want to use the notifications engine for that purpose, because when using the notifications engine, the rollout will proceed in parallel to the notification. However, we would like the rollout to wait until the notification is confirmed to have been sent. For example, after the "setWeight: 100" step completes, we would like to send a Slack notification and then start the subsequent step. By the way, we can already achieve that by using an inline CustomAnalysisTemplate step that sends the Slack notification, but having a custom step is going to make the code clearer, because sending a notification is not really an "analysis".

**Comment by hpvd** [comment]
just a rough idea for a use case of a step plugin with gated-rollout
in combination with the other types of plugins (traffic router and metrics provider)..

Maybe it's possible to use argo rollout not only within kubernetes but to roll out features e.g. also in frontend via toogle/flag activation
Step 1: feature is available  (all services have been deoloyed in a canary way to the right version)
Step 2: in the next step the feature is enabled with a "second run" of canary rollout...

**Comment by aisrael** [comment]
In our case, we often find ourselves deploying multiple `Deployment`s as a single ArgoCD `App`. Having one `Rollout` per `Deployment` would mean we'd now have to find a way to 'synchronize' them together (e.g., abort all rollouts in case any one of them fails an analysis step).

A step plugin might be able to help by gating on some API or signal 'metric'...

**Comment by agaudreault** [comment]
What is left to be done for this feature to go from alpha -> beta

- [ ] Create Official step plugin example with best practice in argoproj-labs
- [ ] Configurable defaults for error, requeue and minRequeue.
- [ ] Health status in ArgoCD should reflect the "cause"
- [ ] ArgoCD UI extensions to add plugin step information
- [ ] Run step plugin e2e on CI
- [ ] Allow to reload config without restarting container
- [ ] Documentation (necessary for alpha)
  - Explain step plugin behavior
  - Explain configuration
  - Example repository
